peer/abstractlist/list.go

```diff
@@ -0,0 +1,620 @@
+	}
+	return errs
+}
+
+// updateOffline must be run under a list lock.
+func (pl *List) updateOffline(updates peer.ListUpdates) error {
+	var errs error
+	for _, id := range updates.Removals {
+		errs = multierr.Append(errs, pl.removeOffline(id))
+	}
+	for _, id := range updates.Additions {
+		errs = multierr.Append(errs, pl.addOffline(id))
+	}
+	return errs
+}
+
+// Start notifies the List that requests will start coming
+func (pl *List) Start() error {
+	return pl.once.Start(pl.start)
+}
+
+func (pl *List) start() error {
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	all := pl.offlinePeerIdentifiers()
+
+	var err error
+	err = multierr.Append(err, pl.updateOffline(peer.ListUpdates{
+		Removals: all,
+	}))
+	err = multierr.Append(err, pl.updateOnline(peer.ListUpdates{
+		Additions: all,
+	}))
+	return err
+}
+
+// Stop notifies the List that requests will stop coming
+func (pl *List) Stop() error {
+	return pl.once.Stop(pl.stop)
+}
+
+func (pl *List) stop() error {
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	all := pl.onlinePeerIdentifiers()
+
+	var err error
+	err = multierr.Append(err, pl.updateOnline(peer.ListUpdates{
+		Removals: all,
+	}))
+	err = multierr.Append(err, pl.updateOffline(peer.ListUpdates{
+		Additions: all,
+	}))
+	return err
+}
+
+// IsRunning returns whether the peer list is running.
+func (pl *List) IsRunning() bool {
+	return pl.once.IsRunning()
+}
+
+// add retains a peer and sets up a facade (a thin proxy for a peer) to receive
+// connection status notifications from the dialer and track pending request
+// counts.
+//
+// add does not add the peer to the list of peers available for choosing (the
+// Implementation).
+// The facade is responsible for adding and removing the peer from the
+// collection of available peers based on connection status notifications.
+//
+// add must be run inside a list lock.
+func (pl *List) add(id peer.Identifier) error {
+	addr := id.Identifier()
+
+	if _, ok := pl.peers[addr]; ok {
+		return peer.ErrPeerAddAlreadyInList(addr)
+	}
+
+	f := newPeerFacade(pl, id)
+	f.boundOnFinish = f.onFinish
+
+	// The transport must not call back before returning.
+	p, err := pl.transport.RetainPeer(id, f)
+	if err != nil {
+		return err
+	}
+
+	f.peer = p
+	pl.peers[addr] = f
+	f.notifyStatusChanged(p)
+
+	return nil
+}
+
+// addOffline must be run under a list lock.
+func (pl *List) addOffline(id peer.Identifier) error {
+	addr := id.Identifier()
+
+	if _, ok := pl.offlinePeers[addr]; ok {
+		return peer.ErrPeerAddAlreadyInList(addr)
+	}
+
+	pl.offlinePeers[addr] = id
+	return nil
+}
+
+// remove releases and forgets a peer.
+//
+// remove must be run under a list lock.
+func (pl *List) remove(id peer.Identifier) error {
+	addr := id.Identifier()
+
+	f, ok := pl.peers[addr]
+	if !ok {
+		return peer.ErrPeerRemoveNotInList(addr)
+	}
+
+	f.remove()
+	delete(pl.peers, addr)
+
+	// The transport must not call back before returning.
+	return pl.transport.ReleasePeer(id, f)
```

Comment:
> No, but it is safe to do so. I think I would prefer to discover a deadlock in tests if a transport tried to send a notification to the subscriber in the retain/release stack, than relax the unnecessary lock, since we want that unlock to occur even through a panic.


> This is true unfortunately. Would you consider the queuing all of the peers to be released, and execute the ReleasePeer in a single goroutine?

Copying comment from the previous thread that Github won't show..

If we want to use `defer` to avoid worrying about panics, we can use inline-functions.

My only concern is that our tests may miss some deadlocks, so would like to see if we can avoid deadlocks by design.

I like the idea that Alex is proposing, can the peer list manage a separate goroutine for notifications. A common pattern we use is to use a non-blocking channel (for notifications), and a slice of "operations". A background goroutine waits on the notification channel, then processes all operations.

In that world, `Retain` or `Release` can trigger status changes whenever they want, and even if they block indefinitely, the peer list will still be readable.

Reply:
Discussing offline.